aws/rust-runtime/aws-types/src/app_name.rs

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! New-type for a configurable app name.
+
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt;
+
+/// App name that can be configured with an AWS SDK client to become part of the user agent string.
+///
+/// This name is used to identify the application in the user agent that gets sent along with requests.
+///
+/// The name may only have alphanumeric characters and any of these characters:
+/// ```text
+/// !#$%&'*+-.^_`|~
+/// ```
+/// Spaces are not allowed.
+///
+/// App names are recommended to be no more than 50 characters.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AppName(Cow<'static, str>);
+
+impl AsRef<str> for AppName {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+impl fmt::Display for AppName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl AppName {
+    /// Creates a new app name.
+    ///
+    /// This will return an `InvalidAppName` error if the given name doesn't meet the
+    /// character requirements. See [`AppName`] for details on these requirements.
+    pub fn new(app_name: impl Into<Cow<'static, str>>) -> Result<Self, InvalidAppName> {
+        let app_name = app_name.into();
+
+        fn valid_character(c: char) -> bool {
+            match c {
+                _ if c.is_ascii_alphanumeric() => true,
+                '!' | '#' | '$' | '%' | '&' | '\'' | '*' | '+' | '-' | '.' | '^' | '_' | '`'
+                | '|' | '~' => true,
+                _ => false,
+            }
+        }
+        if !app_name.chars().all(valid_character) {
+            return Err(InvalidAppName);
+        }
+        if app_name.len() > 50 {
+            tracing::warn!(
+                "The `app_name` set when configuring the SDK client is recommended \
+                 to have no more than 50 characters."
+            )
```

Comment:
I _think_ this is OK, but we should probably ensure this isn't logspam and only gets logged once in practice (not, eg. on every request)